src/parse/mod.rs

```diff
@@ -126,37 +129,43 @@ impl<I: Iterator<Item = Lexeme>> Iterator for Parser<I> {
     /// | declaration_specifiers declarator compound_statement
     /// ;
     fn next(&mut self) -> Option<Self::Item> {
-        let next = self.pending.pop_front().or_else(|| {
-            while let Some(locatable) = self.match_next(&Token::Semicolon) {
-                warn("extraneous semicolon at top level", locatable.location);
-            }
+        let next = self
+            .pending
+            .pop_front()
+            .map(Result::Ok)
+            .or_else(|| self.error_handler.pop_err().map(Result::Err))
+            .or_else(|| {
+                // Parse more of our file
 
-            // check for end of file
-            if self.peek_token().is_none() {
-                if self.scope.is_global() {
-                    self.leave_scope(self.last_location);
+                // Remove extra semicolons
+                while let Some(locatable) = self.match_next(&Token::Semicolon) {
+                    warn("extraneous semicolon at top level", locatable.location);
                 }
-                return self.pending.pop_front();
-            }
-            let mut decls = match self.declaration() {
-                Ok(decls) => decls,
-                // output errors in program order
-                Err(err) => {
-                    if self.pending.is_empty() {
-                        return Some(Err(err.into()));
+
+                // Check for end of file
+                if self.peek_token().is_none() {
+                    if self.scope.is_global() {
+                        self.leave_scope(self.last_location);
+                    }
+
+                    if self.pending.is_empty() && self.error_handler.is_empty() {
```

Comment:
The check to self.pending is redundant, you already checked it above. `self.leave_scope()` can only cause errors and there's no more tokens left in the stream anyway.

```suggestion
                    if self.error_handler.is_empty() {
```